src/Std/Data/DHashMap/Internal/RawLemmas.lean

```diff
@@ -3065,6 +3087,702 @@ end Const
+    (m.filterMap (fun k => Option.guard (fun v => f k v))).1.Equiv (m.filter f).1 := by
+  rw [filterMap_eq_filter]
+  exact ⟨.rfl⟩
+
+theorem toList_filter
+    {f : (a : α) → β a → Bool} :
+    (m.filter f).1.toList.Perm (m.1.toList.filter (fun p => f p.1 p.2)) := by
+  simp_to_model [filter, toList, Equiv] using List.Perm.rfl
+
+theorem toList_filter_key
+    {f : α → Bool} :
+    (m.filter fun a _ => f a).1.toList.Perm (m.1.toList.filter (fun p => f p.1)) := by
+  simp [toList_filter]
+
+theorem keys_filter_key {f : α → Bool} :
+    (m.filter fun k _ => f k).1.keys.Perm (m.1.keys.filter f) := by
+  simp_to_model [keys, filter]
+  simp only [List.keys_eq_map, List.filter_map, Function.comp_def, List.Perm.rfl]
+
+end raw
+
+theorem isEmpty_filter_iff [LawfulBEq α]
+    {f : (a : α) → β a → Bool} (h : m.1.WF) :
+    (m.filter f).1.isEmpty = true ↔
+      ∀ (k : α) (h : m.contains k = true), f k (m.get k h) = false := by
+  simp_to_model [filter, contains, get, isEmpty] using List.isEmpty_filter_eq_true
+
+theorem isEmpty_filter_eq_false_iff [LawfulBEq α]
+    {f : (a : α) → β a → Bool} (h : m.1.WF) :
+    (m.filter f).1.isEmpty = false ↔
+      ∃ (k : α) (h : m.contains k = true), f k (m.get k h) = true := by
+  simp_to_model [filter, contains, get, isEmpty] using List.isEmpty_filter_eq_false
+
+theorem isEmpty_filter_key_iff [EquivBEq α] [LawfulHashable α]
+    {f : α → Bool} (h : m.1.WF) :
+    (m.filter (fun a _ => f a)).1.isEmpty ↔
+      ∀ (k : α) (h : m.contains k), f (m.getKey k h) = false := by
+  simp_to_model [filter, contains, getKey, isEmpty] using List.isEmpty_filter_key_iff
+
+theorem isEmpty_filter_key_eq_false_iff [EquivBEq α] [LawfulHashable α]
+    {f : α → Bool} (h : m.1.WF) :
+    (m.filter (fun a _ => f a)).1.isEmpty = false ↔
+      ∃ (k : α) (h : m.contains k = true), f (m.getKey k h) := by
+  rw [← Bool.not_eq_true, isEmpty_filter_key_iff m h]
+  simp only [Classical.not_forall, Bool.not_eq_false]
+
+theorem contains_filter [LawfulBEq α]
+    {f : (a : α) → β a → Bool} {k : α} (h : m.1.WF) :
+    (m.filter f).contains k = (m.get? k).any (f k) := by
+  simp_to_model [filter, contains, get?] using List.containsKey_filter
+
+theorem contains_filter_key_iff [EquivBEq α] [LawfulHashable α]
+    {f : α → Bool} {k : α} (h : m.1.WF) :
+    (m.filter (fun a _ => f a)).contains k ↔ ∃ h : m.contains k, f (m.getKey k h) := by
+  simp_to_model [filter, contains, getKey] using List.containsKey_filter_key
+
+theorem contains_of_contains_filter [EquivBEq α] [LawfulHashable α]
+    {f : (a : α) → β a → Bool} {k : α} (h : m.1.WF) :
+    (m.filter f).contains k = true → m.contains k = true := by
+  simp_to_model [filter, contains] using containsKey_of_containsKey_filter
+
+theorem size_filter_le_size [EquivBEq α] [LawfulHashable α]
+    {f : (a : α) → β a → Bool} (h : m.1.WF) :
+    (m.filter f).1.size ≤ m.1.size := by
+  simp_to_model [filter, size] using List.length_filter_le
+
+theorem size_filter_eq_size_iff [LawfulBEq α]
```

Comment:
I think there should be a strengthened version of this (and the variants) whose LHS is about equivalence.